### PR TITLE
Template for setting up an emitter

### PR DESCRIPTION
**Pull request by timotheeguerin**
## Add a new template for scaffolding an emitter

`emitter-ts` template setup the following:
- basic emitter files
- typescript
- test with node test runner with basic test and test host setup
- prettier
- linting with eslint 

## Change to the init area

- Added a new `--template` cli option to allow selecting the template without a prompt.
- Refactor to make it easier to test. **Note that none of the API is exposed yet so its just for internal use.**

## Added e2e test for templates

Scafold the template and then run commands like `npm install`, `npm run build`, `npm run test`, etc. to make sure everything is done correctly. 

This is quite costly so might be worth separating in a different step

### CODE REVIEW

File: packages/compiler/tsconfig.json [hunk 1]
```diff
@@ -6,5 +6,6 @@
     "tsBuildInfoFile": "temp/tsconfig.tsbuildinfo",
     "types": ["node", "mocha"]
   },
-  "include": ["src/**/*.ts", "test/**/*.ts"]
+  "include": ["src/**/*.ts", "test/**/*.ts"],
+  "exclude": ["src/init/templates/*/*"]
```
Comment (timotheeguerin) [inline packages/compiler/tsconfig.json:10]: Todo do we want to move those templates to a dedicated package maybe?

Comment (timotheeguerin) [reply packages/compiler/tsconfig.json:10]: I think that would be cleaner specially if we want to add some e2e test to make sure the templates generate correclty.

**Review by tjprescott — CHANGES_REQUESTED** [review]
Happy to get this in. Can we update the doc for getting started with an emitter to reference the steps necessary?

**Review by tjprescott — APPROVED** [review]

File: packages/compiler/templates/emitter-ts/src/emitter.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { EmitContext, emitFile, resolvePath } from "@typespec/compiler";
+
+export async function $onEmit(context: EmitContext) {
+  if (!context.program.compilerOptions.noEmit) {
```
Comment (bterlson) [inline packages/compiler/templates/emitter-ts/src/emitter.ts:4]: Might it be worth putting something super simple here with the emitter framework, like just printing out model names or some such?

File: packages/compiler/templates/emitter-ts/test/test-host.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import { Diagnostic, resolvePath } from "@typespec/compiler";
+import {
+  createTestHost,
+  createTestWrapper,
+  expectDiagnosticEmpty,
+} from "@typespec/compiler/testing";
+import { TestLibrary } from "../src/testing/index.js";
+
+export async function createMyTestHost() {
```
Comment (bterlson) [inline packages/compiler/templates/emitter-ts/test/test-host.ts:9]: I find these names to be a little awkward with "my" in them, could maybe template in the emitter name in place of "my"?

**Review by bterlson — APPROVED** [review]

File: packages/compiler/templates/emitter-ts/test/test-host.ts (hunk 3 above)
Comment (timotheeguerin) [reply packages/compiler/templates/emitter-ts/test/test-host.ts:9]: Yeah that would be great but I think our template engine right now doesn’t have this capability to change casing. I can do a follow up pr that does

File: packages/compiler/templates/emitter-ts/src/emitter.ts (hunk 2 above)
Comment (timotheeguerin) [reply packages/compiler/templates/emitter-ts/src/emitter.ts:4]: I wasn't sure I really wanted people to try to use the emitter framework today as there is still a lot of issues with it.

I might take a stab during the holidays at solving all those issues Travis and I got when using it and after I think we could setup the boilperplate for the emitter framework usage.

**Pull request merged**